File: com/uva/custom_random/test_random.py
import sys
from sets import Set

# import CustomRandom
import custom_random

class Edge(object):
    def __init__(self, first, second):
        self.first = first
        self.second = second
    def __lt__(self, other):
        return self.first < other.first or (self.first == other.first and self.second < other.second)
    def __eq__(self, other):
        return self.first == other.first and self.second == other.second
    def __hash__(self):
        return self.first + self.second


rgen = custom_random.CustomRandom(42)

sys.stdout.write("gamma(1.0, 1.0):\n")
a = rgen.gamma(1.0, 1.0, 2, 4)
for r in a:
    for c in r:
        sys.stdout.write("%.12f " % c)
    sys.stdout.write("\n")

sys.stdout.write("randn:\n")
a = rgen.randn(2, 4)
for r in a:
    for c in r:
        sys.stdout.write("%.12f " % c)
    sys.stdout.write("\n")

sys.stdout.write("randint(2**20):\n")
for i in xrange(64):
    sys.stdout.write("%d " % rgen.randint(0, 2**20))
    if (i + 1) % 8 == 0:
        sys.stdout.write("\n");
sys.stdout.write("\n");

sys.stdout.write("Create random graph:\n")
graph = Set()
for i in xrange(64):
    a = rgen.randint(0, 1024)
    b = rgen.randint(0, 1024)
    graph.add(Edge(a, b))
for e in sorted(graph):
    sys.stdout.write("(%d,%d) " % (e.first, e.second))
sys.stdout.write("\n");

sys.stdout.write("Sample subgraph:\n");
subgraph = rgen.sample(graph, 32)
for e in sorted(subgraph):
    sys.stdout.write("(%d,%d) " % (e.first, e.second))
sys.stdout.write("\n");

File: c++/test/random/test_random.cc
#include <iostream>
#include <iomanip>
#include <set>

#include <mcmc/random.h>

struct Edge {
	Edge(int first, int second) : first(first), second(second) {
	}

	bool operator== (const Edge &peer) const {
		return first == peer.first && second == peer.second;
	}

	bool operator< (const Edge &peer) const {
		return first < peer.first || (first == peer.first && second < peer.second);
	}

	int		first;
	int		second;
};

int main(int argc, char *argv[]) {
	std::cout << std::fixed << std::setprecision(12);

	auto rgen = mcmc::Random::Random(42);

	std::cout << "gamma(1.0, 1.0):" << std::endl;
	auto ag = rgen.gamma(1.0, 1.0, 2, 4);
	for (auto & r : ag) {
		for (auto & c : r) {
			std::cout << c << " ";
		}
		std::cout << std::endl;
	}

	std::cout << "randn:" << std::endl;
	auto an = rgen.randn(2, 4);
	for (auto & r : an) {
		for (auto & c : r) {
			std::cout << c << " ";
		}
		std::cout << std::endl;
	}
	
	std::cout << "randint(3LL**30):" << std::endl;
	for (int i = 0; i < 64; ++i) {
		std::cout << rgen.randint(0, 3LL << 30) << " ";
		if ((i + 1) % 8 == 0) {
			std::cout << std::endl;
		}
	}
	std::cout << std::endl;

	std::cout << "Create random graph:" << std::endl;
	std::set<Edge> graph;
	for (int i = 0; i < 64; i++) {
		auto a = rgen.randint(0, 1024);
	   	auto b = rgen.randint(0, 1024);
		graph.insert(Edge(a, b));
	}
	for (auto & e : graph) {
		std::cout << "(" << e.first << "," << e.second << ") ";
	}
	std::cout << std::endl;

	std::cout << "Sample subgraph:" << std::endl;
	auto subgraph = rgen.sample(graph, 32);
	for (auto & e : *subgraph) {
		std::cout << "(" << e.first << "," << e.second << ") ";
	}
	std::cout << std::endl;
	delete subgraph;

	return 0;
}
